docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,300 @@
+    localityAwareness:
+      inZone: []
+      crossZone:
+        - groups:
+          - zones: ["zone-1"]
+        fallback: Fail
+```
+
+![Use case 6](assets/031/use_case_6_1.png)
+![Use case 6](assets/031/use_case_6_2.png)
+![Use case 6](assets/031/use_case_6_3.png)
+
+TODO What if i have lots of zones and don’t want to write all of them during the incident. I would like to just exclude single zone, do we want to add something like:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: []
+      crossZone:
+        - exclude:
+            zones: ["zone-3"]
+```
+
+#### Load balance equally to local and zone
+
+![Use case 7](assets/031/use_case_7.png)
+
+We have the couple of use cases for this, but no config yet:
+- You are migrating from universal to k8s, both zones are in the same AZ and I would like to move traffic gradually from one to another
+- You have have some zone with testing or preview environment and you would like to handle live traffic on it for testing. Is it a use case for locality aware lb? Or should this be done with MeshHTTPRoute?
+The question is if we really need it, or can this be done currently in some other way?
+
+
+### Implementation
+
+We want to enhance current `MeshLoadBalancingStrategy` so code related to the changes should be placed in the plugin.
+
+#### Matching endpoints
+Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.
+
+Example:
+
+On Kubernetes, all tags from the pod are populated automatically, while on Universal, users need to populate tags themselves:
+
+```yaml
+type: Dataplane
+mesh: default
+name: demo
+networking:
+  address: 127.0.0.1
+  inbound:
+    - port: 9999
+      servicePort: 10000
+      serviceAddress: 127.0.0.1
+      tags:
+        kuma.io/service: demo
+        kuma.io/protocol: tcp
+        k8s.io/az: us-east-1
+```
+
+
+Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::
+
+```json
+"metadata":{
+  "filter_metadata": {
+    "envoy.lb": {
+      "az/test": "b",
+      "kuma.io/protocol": "tcp",
+      "k8s.io/az": "us-east-1"
+    }
+  }
+}
+```
+
+
+We can retrieve them from `ClusterLoadAssignment` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.
+
+Pseudo algorithm:
+
+1. Get all inbound tags.
+2. Take the matched rules.
+3. Check if inbound has the tag defined in the rule. 
+4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule( first inZone, then  cross Zone). If no, get all the endpoints in the zone. 
```

Comment:
So by default cross Zone traffic would be off? I think it's an ok behaviour especially as this is compliant with regular k8s behaviour

Reply:
I think we really need to cover details around provision factor etc here. Especially if there are endpoints we could technically route to that we don't send down to Envoy (I'm not sure this can happen though :) ).

Reply:
Do you mean Envoy's provision factor or maybe how we are going to choose endpoints to ensure there are endpoints?